boot/mbed/src/flash_map_backend.cpp

```diff
@@ -0,0 +1,211 @@
+#ifndef MCUBOOT_OVERWRITE_ONLY
+        (mbed::BlockDevice*) &mcuboot_scratch_bd        /** Scratch space for swapping images */
+#else
+        nullptr
+#endif
+};
+
+static struct flash_area flash_areas[FLASH_AREAS];
+
+int flash_area_open(uint8_t id, const struct flash_area** fapp) {
+
+    *fapp = &flash_areas[id];
+    struct flash_area* fap = (struct flash_area*)*fapp;
+
+    // The offset of the slot is from the beginning of the flash device.
+    switch (id) {
+        case PRIMARY_ID:
+            fap->fa_off = MCUBOOT_PRIMARY_SLOT_START_ADDR;
+            break;
+        case SECONDARY_ID:
+            // The offset of the secondary slot is not currently used.
+            fap->fa_off = 0;
+            break;
+#ifndef MCUBOOT_OVERWRITE_ONLY
+        case SCRATCH_ID:
+            fap->fa_off = MCUBOOT_SCRATCH_START_ADDR;
+            break;
+#endif
+        default:
+            MCUBOOT_LOG_ERR("flash_area_open, unknown id %d", id);
+            return -1;
+    }
+
+    fap->fa_id = id;
+    fap->fa_device_id = 0; // not relevant
+
+    mbed::BlockDevice* bd = flash_map_bd[id];
+    fap->fa_size = (uint32_t) bd->size();
+    return bd->init();
+}
+
+void flash_area_close(const struct flash_area* fap) {
+    mbed::BlockDevice* bd = flash_map_bd[fap->fa_id];
+    bd->deinit();
+}
+
+/*
+ * Read/write/erase. Offset is relative from beginning of flash area.
+ */
+int flash_area_read(const struct flash_area* fap, uint32_t off, void* dst, uint32_t len) {
+    mbed::BlockDevice* bd = flash_map_bd[fap->fa_id];
+
+    // Note: The address must be aligned to bd->get_read_size(). If MCUBOOT_READ_GRANULARITY
+    // is defined, the length does not need to be aligned.
+#ifdef MCUBOOT_READ_GRANULARITY
+    uint32_t read_size = bd->get_read_size();
+    if (read_size == 0) {
+        MCUBOOT_LOG_ERR("Invalid read size: must be non-zero");
+        return -1;
+    }
+    if (MCUBOOT_READ_GRANULARITY < read_size) {
+        MCUBOOT_LOG_ERR("Please increase MCUBOOT_READ_GRANULARITY (currently %u) to be at least %u",
+                    MCUBOOT_READ_GRANULARITY, read_size);
+        return -1;
+    }
+
+    uint32_t remainder = len % read_size;
+    len -= remainder;
+    if (len != 0) {
+#endif
+        if (!bd->is_valid_read(off, len)) {
+            MCUBOOT_LOG_ERR("Invalid read: fa_id %d offset 0x%x len 0x%x", fap->fa_id, off, len);
+            return -1;
+        }
+        else {
+            int ret = bd->read(dst, off, len);
+            if (ret != 0) {
+                MCUBOOT_LOG_ERR("Read failed: fa_id %d offset 0x%x len 0x%x", fap->fa_id, off, len);
+                return ret;
+            }
+        }
+#ifdef MCUBOOT_READ_GRANULARITY
+    }
+
+    if (remainder) {
+        if (!bd->is_valid_read(off + len, read_size)) {
+            MCUBOOT_LOG_ERR("Invalid read: fa_id %d offset 0x%x len 0x%x", fap->fa_id, off + len, read_size);
+            return -1;
+        }
+        else {
+            uint8_t buffer[read_size];
```

Comment:
We initially used a predefined array size, because Mbed OS enables `-Wvla` which warns usage of variable-sized arrays. But variable-sized array does bring in flexibility in our case.

@AGlass0fMilk In this case, we can remove the config from `mbed_lib.json` and the `#ifdef`s in this function?

Reply:
With `MCUBOOT_READ_GRANULARITY` being a compile time var, and `read_size` a runtime var it makes sense that one would get warnings, so it's OK to rollback on my initial comment and use `MCUBOOT_READ_GRANULARITY` instead.

Reply:
I'm personally okay with both options. If the static array size (MCUBOOT_READ_GRANULARITY) is used, we can document (in `mbed-app.json`) it should be at least the value of `<blockdevice>->get_read_size()`.

Reply:
We prefer compile time var, hence why Mbed OS enables `-Wvla`. So I prefer the original option with predefined array size.

Reply:
I prefer the static array size implementation. I will update the `mbed_lib.json` help for the `MCUBOOT_READ_GRANULARITY` setting with the above information.